src/state/Maps/selectors.js

```diff
@@ -422,36 +490,240 @@ const getAllLayersStateByMapKey = createCachedSelector(
+
+		// modifiers defined by key
+		let metadataDefinedByKey = layer.metadataModifiers ? {...layer.metadataModifiers} : {};
+
+		// Get actual metadata keys defined by filterByActive
+		const activeMetadataKeys = common.getActiveKeysByFilterByActiveObserver(layer.filterByActive);
+
+		// Merge metadata, metadata defined by key have priority
+		const mergedMetadataKeys = commonHelpers.mergeMetadataKeys(metadataDefinedByKey, activeMetadataKeys);
+
+		// It converts modifiers from metadataKeys: ["A", "B"] to metadataKey: {in: ["A", "B"]}
+		let relationsFilter = commonHelpers.convertModifiersToRequestFriendlyFormat(mergedMetadataKeys);
+
+		// add layerTemplate od areaTreeLevelKey
+		if (layer.layerTemplateKey) {
+			relationsFilter.layerTemplateKey = layer.layerTemplateKey;
+		} else if (layer.areaTreeLevelKey) {
+			relationsFilter.areaTreeLevelKey = layer.areaTreeLevelKey;
+		}
+		return relationsFilter;
+	} else {
+		return null;
+	}
+});
+
+/**
+ * @param layerState {Object}
+ */
+const getAttributeRelationsFilterFromLayerState = createRecomputeSelector((layerState) => {
+    const spatialFilter = getSpatialRelationsFilterFromLayerState(layerState);
+    if(spatialFilter) {
+        const attributeFilter = {...spatialFilter};
+        if(layerState.styleKey) {
+            // add styleKey
+            attributeFilter.styleKey = layerState.styleKey;
         }
+        return attributeFilter;
+    } else {
+        return null;
     }
-)((state, mapKey) => mapKey);
+})
+
+/**
+ * @param spatialDataSource {Object}
+ * @param layerState {Object} layer definition from state or passed to the Map component
+ * @param layerKey {string} layer unique identifier
+ * @param attributeDataSourceKeyAttributeKeyPairs {Object} key-value pairs, where key is attribute data source key and value is matching attribute key
+ * @param mapKey {string} map unique identifier
+ * @param spatialRelationsFilter {Object} see getSpatialRelationsFilterFromLayerState
+ * @param attributeRelationsFilter {Object} see getAttributeRelationsFilterFromLayerState
+ */
+const getFinalLayerByDataSourceAndLayerState = createRecomputeSelector((spatialDataSource, layerState, layerKey, attributeDataSourceKeyAttributeKeyPairs, mapKey, spatialRelationsFilter, attributeRelationsFilter) => {
+	let {attribution, nameInternal, type, fidColumnName, geometryColumnName,  ...dataSourceOptions} = spatialDataSource?.data;
+	let {key, name, opacity, styleKey, renderAsType, options: layerStateOptions} = layerState;
+
+	layerKey = layerKey || key;
+
+	// TODO temporary for development. Next, could be data source type rewritten in layer state (e.g. vector -> tiled-vector?)
+	if (renderAsType) {
+		type = renderAsType;
+	}
+
+	let options = {};
+	if (dataSourceOptions && layerStateOptions) {
+		options = {...dataSourceOptions, ...layerStateOptions};
+	} else if (layerStateOptions) {
+		options = {...layerStateOptions}
+	} else if (dataSourceOptions) {
+		options = {...dataSourceOptions};
+	}
```

Comment:
Are the ifs actually needed here?

Reply:
You're right. This https://github.com/gisat-panther/ptr-state/pull/56/commits/01c74b97608ecc19b43655795852a3f8e21e4aae should be enough

Reply:
Why the null?

Reply:
Fixed cc4bd6b485879d6ba819aa0eadb4e1b248c91ab4